corehq/ex-submodules/casexml/apps/phone/utils.py

```diff
@@ -15,3 +15,35 @@ def delete_sync_logs(before_date, limit=1000, num_tries=10):
             pass
 
     raise CouldNotPruneSyncLogs()
+
+
+ITEMS_COMMENT_PREFIX = b'<!--items='
+
+
+def get_cached_items_with_count(cached_bytes):
+    """Get the number of items encoded in cached XML elements byte string
+
+    The string, if it contains an item count, should be prefixed with
+    b'<!--items=' followed by one or more numeric digits (0-9), then
+    b'-->', and finally the cached XML elements.
+
+    Example: b'<!--items=42--><fixture>...</fixture>'
+
+    If the string does not start with b'<!--items=' then it is assumed
+    to contain only XML elements.
+
+    :returns: Two-tuple: (xml_elements_bytes, num_items)
+    """
+    if cached_bytes.startswith(ITEMS_COMMENT_PREFIX):
+        DIGITS = b'0123456789'
+        digits = []
+        i = len(ITEMS_COMMENT_PREFIX)
+        total_bytes = len(cached_bytes)
+        while i < total_bytes and cached_bytes[i] in DIGITS:
+            digits.append(cached_bytes[i])
+            i += 1
+        if digits and cached_bytes[i:i + 3] == b'-->':
+            num_items = int(b"".join(digits))
+            return cached_bytes[i + 3:], num_items
```

Comment:
Wouldn't it be easier to use a regex to get this out?

Reply:
I could do that, although regex engines are a bit of a black box and I didn't want it to end up doing something unexpected like searching backward from the end of the string. The time characteristics of this code are well defined and bounded for sane inputs.

On the other hand, I just remembered that Python has [`regex.match`](https://docs.python.org/3/library/re.html#re.regex.match) (I always forget that one), which appears to have been made for just this use case. I'll use that instead since it does make the code simpler. 

https://github.com/dimagi/commcare-hq/pull/18290/commits/815a163f63d113004914b673aa5bb5f463ff7487